File: packages/shared/schemas/recruitment-custom-field.ts
import { z } from "zod";

export const CreateRecruitmentCustomFieldSchema = z.object({
  recruitmentId: z.string(),
  recruitment: z.unknown(),
  fieldLabel: z.string(),
  fieldType: z.string(),
  fieldOptions: z.array(z.string()).optional(),
  isRequired: z.boolean(),
  orderIndex: z.number().int(),
  version: z.number().int(),
});

export const UpdateRecruitmentCustomFieldSchema = z.object({
  recruitmentId: z.string().optional(),
  recruitment: z.unknown().optional(),
  fieldLabel: z.string().optional(),
  fieldType: z.string().optional(),
  fieldOptions: z.array(z.string()).optional(),
  isRequired: z.boolean().optional(),
  orderIndex: z.number().int().optional(),
  version: z.number().int().optional(),
});

export type CreateRecruitmentCustomFieldInput = z.infer<typeof CreateRecruitmentCustomFieldSchema>;
export type UpdateRecruitmentCustomFieldInput = z.infer<typeof UpdateRecruitmentCustomFieldSchema>;

File: apps/api/internal/models/recruitment_custom_field.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/datatypes"
	"gorm.io/gorm"
)

// RecruitmentCustomField represents a recruitmentcustomfield in the system.
type RecruitmentCustomField struct {
	ID        string         `gorm:"primarykey;size:36" json:"id"`
	RecruitmentID string `gorm:"size:36;index" json:"recruitment_id" binding:"required"`
	Recruitment Recruitment `gorm:"foreignKey:RecruitmentID" json:"recruitment"`
	FieldLabel string `gorm:"size:255" json:"field_label" binding:"required"`
	FieldType string `gorm:"size:255" json:"field_type" binding:"required"`
	FieldOptions datatypes.JSONSlice[string] `gorm:"type:json" json:"field_options"`
	IsRequired bool `json:"is_required"`
	OrderIndex int `json:"order_index"`
	Version   int            `gorm:"not null;default:1" json:"version"`
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"-"`
}

// BeforeCreate generates a UUID before inserting.
func (m *RecruitmentCustomField) BeforeCreate(tx *gorm.DB) error {
	if m.ID == "" {
		m.ID = uuid.New().String()
	}
	return nil
}

// BeforeUpdate increments Version so offline clients can detect server-side updates.
func (m *RecruitmentCustomField) BeforeUpdate(tx *gorm.DB) error {
	m.Version++
	return nil
}
